src/Features/CSharp/Portable/Completion/CompletionProviders/ObjectAndWithInitializerCompletionProvider.cs

```diff
@@ -151,6 +154,15 @@ internal override bool IsInsertionTrigger(SourceText text, int characterPosition
                 return Tuple.Create(expectedType, token.GetLocation());
             }
 
+            // expr with { $$
+            if (token.Parent.Parent.IsKind(SyntaxKind.WithExpression, out WithExpressionSyntax withExpression))
+            {
+                var type = semanticModel.GetTypeInfo(withExpression.Receiver, cancellationToken).Type;
+                // Note: no special handling for type parameters since they can't be used in 'with' expressions ('with' is restricted to records at the moment)
+
+                return Tuple.Create(type, token.GetLocation());
```

Comment:
what if `.Type` returns `null`?  is the rest of the code resilient.  

do we have a test for a `with` on an expression we don't understand?

Reply:
can we also have a tracking bug since with/object/collection are different rigth?  i.e. in an object initializer we'll show fields/props that a `with` might not.  I think that's fine initially, but we would want to get the set 'right'.

Reply:
>  i.e. in an object initializer we'll show fields/props that a with might not. 

In a `with` initializer, you can't assign to indexers `x with { [0] = 1 }`, but I don't think this difference affects this completion provider.
Is there some other difference I'm missing?

Reply:
> what if .Type returns null? is the rest of the code resilient.

I forced `.Type` to `null` here and then tested manually in VS. No issue.

> do we have a test for a with on an expression we don't understand?

I tested manually for `error with { ... }` and `typeWithoutClone with { ... }`.

Reply:
> Is there some other difference I'm missing?

FOr example, say you had:

```c#
public record Whatever
{
     public List<string> Items { get; }
}
```

You can say:

`var w = new Wahtever { Items = { "Foo" } }`

But you can't say:

`var w1 = w with { Items = ... }`